driver/src/test/java/org/neo4j/driver/internal/cluster/LoadBalancerTest.java

```diff
@@ -18,616 +18,145 @@
+        RoutingTable routingTable = mock( RoutingTable.class );
+        Rediscovery rediscovery = mock( Rediscovery.class );
+        when( routingTable.isStale() ).thenReturn( true );
+        HashSet<BoltServerAddress> set = new HashSet<>( asList( new BoltServerAddress( "abc", 12 ) ) );
+        when( routingTable.update( any( ClusterComposition.class ) ) ).thenReturn( set );
 
         // when
-        routing.acquireReadConnection();
+        LoadBalancer balancer = new LoadBalancer( DEV_NULL_LOGGER, conns, routingTable, rediscovery );
 
         // then
-        events.assertCount( any( ClusterTopology.CompositionRequest.class ), equalTo( 2 ) );
+        InOrder inOrder = inOrder( rediscovery, routingTable, conns );
+        inOrder.verify( rediscovery ).lookupRoutingTable( conns, routingTable );
+        inOrder.verify( routingTable ).update( any( ClusterComposition.class ) );
+        inOrder.verify( conns ).purge( new BoltServerAddress( "abc", 12 ) );
     }
 
-    @Test
-    public void shouldReconnectWithRouterIfNoReadersAreAvailable() throws Exception
-    {
-        // given
-        cluster.on( "one", 1337 )
-                .ttlSeconds( 20 )
-                .provide( "one", 1337, WRITE, ROUTE )
-                .provide( "two", 1337, ROUTE );
-        cluster.on( "two", 1337 )
-                .ttlSeconds( 20 )
-                .provide( "one", 1337, READ, WRITE, ROUTE )
-                .provide( "two", 1337, READ, ROUTE );
-        connections.up( "one", 1337 ).up( "two", 1337 );
-
-        LoadBalancer routing = seedLoadBalancer( "one", 1337 );
-
-        events.assertCount( any( ClusterTopology.CompositionRequest.class ), equalTo( 1 ) );
-
-        cluster.on( "one", 1337 )
-                .ttlSeconds( 20 )
-                .provide( "one", 1337, READ, WRITE, ROUTE )
-                .provide( "two", 1337, READ, ROUTE );
-
-        // when
-        routing.acquireWriteConnection(); // we should require the presence of a READER even though we ask for a WRITER
-
-        // then
-        events.assertCount( any( ClusterTopology.CompositionRequest.class ), equalTo( 2 ) );
-    }
 
     @Test
-    public void shouldReconnectWithRouterIfNoWritersAreAvailable() throws Exception
+    public void shouldEnsureRoutingOnInitialization() throws Exception
     {
-        // given
-        cluster.on( "one", 1337 )
-                .ttlSeconds( 20 )
-                .provide( "one", 1337, READ, ROUTE )
-                .provide( "two", 1337, READ, WRITE, ROUTE );
-        cluster.on( "two", 1337 )
-                .ttlSeconds( 20 )
-                .provide( "one", 1337, READ, ROUTE )
-                .provide( "two", 1337, READ, WRITE, ROUTE );
-        connections.up( "one", 1337 );
-
-        events.registerHandler( StubConnectionPool.EventSink.class, new StubConnectionPool.EventSink.Adapter()
-        {
+        // given & when
+        final AtomicInteger ensureRoutingCounter = new AtomicInteger( 0 );
```

Comment:
What do you think about using `Mockito#spy()` instead of atomic integers and runnables? Invocations could then be verified as with regular mockito mocks.